src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -313,13 +322,52 @@ private void onData(
             }
             else if (payload != null)
             {
+                final long streamId = data.streamId();
                 decodeTraceId = data.trace();
-                DirectBuffer buffer = payload.buffer();
 
+                DirectBuffer buffer = payload.buffer();
                 int offset = payload.offset();
-                int length = payload.sizeof();
-
-                decodeState.decode(buffer, offset, length);
+                int limit = payload.limit();
+
+                if (bufferSlot != NO_SLOT)
+                {
+                    final MutableDirectBuffer slotBuffer = bufferPool.buffer(bufferSlot);
+                    slotBuffer.putBytes(slotLimit, buffer, offset, limit - offset);
+                    slotLimit += limit - offset;
+                    buffer = slotBuffer;
+                    offset = 0;
+                    limit = slotLimit;
+                }
+
+                while (offset < limit)
+                {
+                    final MqttPacketFixedHeaderFW packet = mqttPacketFixedHeaderRO.tryWrap(buffer, offset, limit);
+
+                    if (packet == null || packet.limit() + packet.remainingLength() > limit)
+                    {
+                        break;
+                    }
+
+                    final MqttPacketType packetType = MqttPacketType.valueOf(packet.typeAndFlags() >> 4);
+                    offset += decodeState.decode(packetType, buffer, offset, limit);
+                }
+
+                if (offset < limit)
+                {
+                    if (bufferSlot == NO_SLOT)
+                    {
+                        bufferSlot = bufferPool.acquire(streamId);
```

Comment:
Rename `bufferSlot` to `slotIndex` so that all slot-related variables start with `slot` prefix for readability.